Moved XIP expansion to a temporal directory

juanjonol opened the pull request:
Instead of expanding XIPs on its current directory, a temporal directory is used. The main advantage of this approach is that, because the temporal directory is created on the same volume where Xcode will be installed, the installation is vastly improved if the XIP is on a different volume.

- It also makes more conceptual sense to keep a temporal file on the OS temporal folders.
- https://nshipster.com/temporary-files/
- The implementation of the `temporalDirectory` function and variable are based on the similar `trashItem`, but without the `@discardableResult` annotation (I cannot think of any case where this URL could to be ignored).

This closes #178.

___

I've tested this on September 21st, to get some hard data about the improvement with this PR.

Installing Xcode 14.1 beta 2 (14B5024i) on a different volume that the one where the .xip is located, the `(3/6) Moving Xcode to /Volumes/<path>/Xcode-14-1-0-Beta.2.app` phase with the old implementation (using the `--xip-expand-inplace` flag) takes 4.78 minutes on a 2021 MacBook Pro 14". Without `--xip-expand-inplace`, this is virtually instantaneous (0.0009 seconds).

For comparison, the improvement I get using `--experimental-unxip` over regular unxip is 1.33 minutes.

juanjonol committed:
Moved XIP expansion to a temporal directory

Instead of expanding XIPs on its current directory, a temporal directory is used. The main advantage of this approach is that, because the temporal directory is created on the same volume where Xcode will be installed, the installation is vastly improved if the XIP is on a different volume.

- https://nshipster.com/temporary-files/
- The implementation of the `temporalDirectory` function and variable are based on the similar `trashItem`, but without the `@discardableResult` annotation (I cannot think of any case where this URL could to be ignored).

This closes #178.

juanjonol commented:
A few comments I didn't add to the commit:

- This always changes the expansion directory, not just when the destination volume is different that the one where the XIP is. This means the XIP will never be expanded on `~/Library/Application Support/com.robotsandpencils.xcodes/` again.[^1] I did this for consistency and because I don't think there is anything useful that could be done with a partially-expanded Xcode version. This also allows macOS to automatically delete the expanded data if for some reason `xcodes` fails to move it.
- @MattKiazyk suggested [a change similar to other one on Xcodes.app](https://github.com/RobotsAndPencils/XcodesApp/pull/155/files), but I'm not sure if it makes sense here, for the same reasons as the previous point.

[^1]: Theoretically, `url(for:in:appropriateFor:create:)` can fail, and in that case the original code will be used, but I haven't found any case where it actually fails.

MattKiazyk commented on Sources/XcodesKit/XcodeInstaller.swift:
```diff
@@ -708,9 +708,10 @@ public final class XcodeInstaller {
     }
 
     func unarchiveAndMoveXIP(at source: URL, to destination: URL) -> Promise<URL> {
+        let xcodeExpansionDirectory = (try? Current.files.temporalDirectory(for: source)) ?? source.deletingLastPathComponent()
```
#picky Can we move the `(try? Current.files.temporalDirectory(for: source)) ?? source.deletingLastPathComponent()` into it's own function inside of `Files` so that we can just call `try? Current.files.xcodeExpansionDirectory`

juanjonol replied:
I tried to solve this on https://github.com/RobotsAndPencils/xcodes/pull/179/commits/f1163aaf45111dcca1bc9a664fd8f49e7240ee3b. Please take a look! And don't worry about being picky 😊

juanjonol replied:
I mark this as resolved to remove some noise from this Pull Request, please fill free to reopen it if you have some comments!

MattKiazyk commented:
@juanjonol Thanks for the PR! It is appreciated! I love what you did here. I have one picky comment plus another ask.

Can we add a flag onto the cli so that the user could (if they wanted) to NOT expand to the temporary directory and instead to where it was downloaded? I agree temporary is better, but just wanted to give that option to the user if they wanted to revert back to the old behaviour

Thanks!

juanjonol committed:
Moved logic to get the Xcode's expansion directory to a function inside `Files`

As suggested by @MattKiazyk, to be able to get the expansion directory by simply calling `Current.files.xcodeExpansionDirectory()`.

- The `Files.temporalDirectory()` added on the previous commits is keep even if it's not used outside `Files.xcodeExpansionDirectory()`, to keep each function with a single statement (as the other functions on `Files`).
- I'm not sure there's value on adding the `xcodeExpansionDirectory` variable, but added for consistency with the other functions on `Files`.

juanjonol committed:
Added `expand-xip-inplace` flag to prevent expanding on a temporal directory

As suggested by @MattKiazyk, to allow the user to revert to the old behaviour if needed.

- For testing this flag is set to `true`, to use a well-known location when testing.
- The `xcodeExpansionDirectory` function cannot be easily expressed on a single line anymore, so removed the variable added on the previous commit (which wasn’t used anyway).

juanjonol commented:
@MattKiazyk I added a `expand-xip-inplace` flag to keep the old behaviour (simply using the directory where the .xip is). If you need anything else or you don't like how it's named/implemented, don't hesitate to tell me (I'm picky too!).

Also, I though about adding tests, but I don't know how to test this. It's a small change so maybe it isn't needed, but if you have any idea about it, we can look into it.

juanjonol commented:
Hi @MattKiazyk, this PR has been stale for a few months. Could you please review it? I see there're conflicts now, but there's no point fixing them if the PR won't be merged...

juanjonol committed:
Merge tag '0.20.0' into xip-multivolume-expasion

# Conflicts:
#	Sources/XcodesKit/XcodeInstaller.swift
#	Sources/xcodes/main.swift

juanjonol committed:
Merge tag '1.0.0' into xip-multivolume-expasion

# Conflicts:
#	Sources/XcodesKit/XcodeInstaller.swift
#	Sources/xcodes/main.swift
#	Tests/XcodesKitTests/XcodesKitTests.swift

juanjonol commented:
I've updated this PR with the latest changes on main. All test are green and it can be merged without conflicts.

@MattKiazyk could you please take a look at this and tell me if there's something preventing its merge? This still says that there're changes requested, but I made those changes months ago...

juanjonol committed:
Merge tag '1.1.0' into xip-multivolume-expasion

# Conflicts:
#	Sources/XcodesKit/XcodeInstaller.swift
#	Sources/xcodes/App.swift

juanjonol committed:
Merge tag '1.2.0' into xip-multivolume-expasion

# Conflicts:
#	Sources/XcodesKit/Environment.swift
#	Sources/XcodesKit/XcodeInstaller.swift
#	Sources/xcodes/App.swift
#	Tests/XcodesKitTests/Environment+Mock.swift
#	Tests/XcodesKitTests/XcodesKitTests.swift

juanjonol committed:
Merge tag '1.3.0' into xip-multivolume-expasion

# Conflicts:
#	Sources/xcodes/App.swift

juanjonol committed:
Merge tag '1.4.1' into xip-multivolume-expasion

1.4.1

juanjonol committed:
Merge tag '1.5.0' into xip-multivolume-expasion

1.5.0

juanjonol committed:
Merge tag '1.6.0' into xip-multivolume-expasion

1.6.0

juanjonol committed:
Merge tag '1.6.2' into xip-multivolume-expasion

1.6.2

juanjonol committed:
Merge tag '2.0.1' into xip-multivolume-expasion

2.0.1

# Conflicts:
#	Sources/XcodesKit/Environment.swift
#	Sources/XcodesKit/XcodeInstaller.swift
#	Sources/xcodes/App.swift
#	Tests/XcodesKitTests/Environment+Mock.swift
#	Tests/XcodesKitTests/XcodesKitTests.swift

juanjonol committed:
Merge tag '2.0.2' into xip-multivolume-expasion

2.0.2
